snmalloc-edp/tests/global_alloc.rs

```diff
@@ -0,0 +1,111 @@
+use std::{alloc::{self, GlobalAlloc}, cell::Cell, ptr};
+
+use snmalloc_edp::*;
+
+thread_local! {
+    static THREAD_ALLOC: Cell<*mut Alloc> = const { Cell::new(ptr::null_mut()) };
+}
+
+#[no_mangle]
+pub fn __rust_get_thread_allocator() -> *mut Alloc {
+    THREAD_ALLOC.get()
+}
+
+struct System;
+
+unsafe impl alloc::GlobalAlloc for System {
+    #[inline]
+    unsafe fn alloc(&self, layout: alloc::Layout) -> *mut u8 {
+        // SAFETY: the caller must uphold the safety contract for `malloc`
+        sn_rust_alloc(layout.align(), layout.size())
+    }
+
+    #[inline]
+    unsafe fn alloc_zeroed(&self, layout: alloc::Layout) -> *mut u8 {
+        // SAFETY: the caller must uphold the safety contract for `malloc`
+        sn_rust_alloc_zeroed(layout.align(), layout.size())
+    }
+
+    #[inline]
+    unsafe fn dealloc(&self, ptr: *mut u8, layout: alloc::Layout) {
+        // SAFETY: the caller must uphold the safety contract for `malloc`
+        sn_rust_dealloc(ptr, layout.align(), layout.size())
+    }
+
+    #[inline]
+    unsafe fn realloc(&self, ptr: *mut u8, layout: alloc::Layout, new_size: usize) -> *mut u8 {
+        // SAFETY: the caller must uphold the safety contract for `malloc`
+        sn_rust_realloc(ptr, layout.align(), layout.size(), new_size)
+    }
+}
+
+// SAFETY: this should only be called once per thread, and the global
+// allocator shouldn't be used outside of this function
+unsafe fn with_thread_allocator<F: FnOnce() -> R, R>(f: F) -> R {
+    unsafe {
+        let layout = alloc::Layout::from_size_align(sn_alloc_size, sn_alloc_align).unwrap();
+        // TODO: bootstrap the thread-local allocator allocation a different way
```

Comment:
TODO

Reply:
Is this todo still relevant? Because with @raoulstrackx we have figured that the way to use `sn_thread_init(allocator);`
and `sn_thread_cleanup(allocator);` is to put them around https://github.com/rust-lang/rust/blob/master/library/std/src/sys/pal/sgx/thread.rs#L43

Reply:
> put them around https://github.com/rust-lang/rust/blob/master/library/std/src/sys/pal/sgx/thread.rs#L43.

That is not right. That code is only run for additional threads besides the main thread. It should be somewhere in entry.

> Is this todo still relevant?

Yes you still need to allocate memory for the thread allocator.